File: agent/compose.ts
import Anthropic from "@anthropic-ai/sdk";
import express, { Request, Response } from "express";
import dotenv from "dotenv";
import { z } from "zod";

dotenv.config();

// ── Anthropic client ──────────────────────────────────────────────────────────

const anthropic = new Anthropic({
  apiKey: process.env.ANTHROPIC_API_KEY,
});

// ── Zod schemas ───────────────────────────────────────────────────────────────

const CatalogueAssetSchema = z.object({
  address: z.string(),
  symbol: z.string(),
  name: z.string(),
  sector: z.string(),
  currentPriceUsdg: z.string().optional(),
  isActive: z.boolean().optional(),
});

const ProposedConstituentSchema = z.object({
  address: z.string(),
  symbol: z.string(),
  weightBps: z.number().int().min(100).max(5000),
  rationale: z.string().min(1),
});

const AIProposalSchema = z.object({
  constituents: z
    .array(ProposedConstituentSchema)
    .min(3)
    .max(12),
  overallRationale: z.string().min(1),
  riskNotes: z.string().min(1),
});

type CatalogueAsset = z.infer<typeof CatalogueAssetSchema>;
type AIProposal = z.infer<typeof AIProposalSchema>;

const ComposeRequestSchema = z.object({
  thesis: z.string().min(20).max(2000),
});

// ── System prompt ─────────────────────────────────────────────────────────────

function buildSystemPrompt(): string {
  return `You are a financial analyst building thematic equity baskets from a specific catalogue of tokenized stocks. You will receive a natural language investment thesis and a JSON catalogue of available stocks with their symbols, names, sectors, and current prices.

Your task is to select 3 to 12 stocks from the catalogue that best represent the thesis, assign each a target weight in basis points that sum to exactly 10,000, and provide a one-sentence rationale for each inclusion.

Rules:
- Only select stocks that appear in the provided catalogue
- Weights must be integers and must sum to exactly 10,000
- No single weight may be less than 100 (1%) or more than 5000 (50%)
- Do not include more than 12 constituents
- Do not include fewer than 3 constituents
- Prefer direct plays over indirect beneficiaries unless the thesis specifically calls for breadth
- Weight by conviction and relevance to the thesis, not by market cap alone

Return ONLY valid JSON matching this schema, no preamble or explanation outside the JSON:
{
  "constituents": [
    {
      "address": "0x...",
      "symbol": "AAPL",
      "weightBps": 2000,
      "rationale": "one sentence explaining why this stock fits the thesis"
    }
  ],
  "overallRationale": "two to three sentences explaining the basket overall construction logic",
  "riskNotes": "one to two sentences noting the key risks or concentration exposures"
}`;
}

function buildUserPrompt(thesis: string, catalogue: CatalogueAsset[]): string {
  const catalogueJSON = JSON.stringify(
    catalogue.map((a) => ({
      address: a.address,
      symbol: a.symbol,
      name: a.name,
      sector: a.sector,
      currentPriceUsdg: a.currentPriceUsdg ?? "unknown",
    })),
    null,
    2
  );

  return `Investment thesis: ${thesis}

Available stock catalogue:
${catalogueJSON}

Return only the JSON proposal with no additional text.`;
}

// ── Validation ────────────────────────────────────────────────────────────────

function validateProposal(
  raw: unknown,
  catalogue: CatalogueAsset[]
): { valid: true; proposal: AIProposal } | { valid: false; error: string } {
  const parsed = AIProposalSchema.safeParse(raw);
  if (!parsed.success) {
    return {
      valid: false,
      error: `Schema validation failed: ${parsed.error.message}`,
    };
  }

  const proposal = parsed.data;
  const catalogueAddresses = new Set(catalogue.map((a) => a.address.toLowerCase()));
  const catalogueByAddress = new Map(catalogue.map((a) => [a.address.toLowerCase(), a]));

  // Verify all proposed addresses exist in the catalogue.
  for (const c of proposal.constituents) {
    if (!catalogueAddresses.has(c.address.toLowerCase())) {
      return {
        valid: false,
        error: `Constituent address ${c.address} (${c.symbol}) not found in catalogue`,
      };
    }

    // Cross-check symbol matches the catalogue entry for that address.
    const catalogueEntry = catalogueByAddress.get(c.address.toLowerCase());
    if (catalogueEntry && catalogueEntry.symbol !== c.symbol) {
      return {
        valid: false,
        error: `Symbol mismatch: address ${c.address} is ${catalogueEntry.symbol} in catalogue but proposal says ${c.symbol}`,
      };
    }
  }

  // Verify no duplicate addresses.
  const addresses = proposal.constituents.map((c) => c.address.toLowerCase());
  const uniqueAddresses = new Set(addresses);
  if (uniqueAddresses.size !== addresses.length) {
    return { valid: false, error: "Duplicate constituent addresses in proposal" };
  }

  // Verify weight sum is exactly 10,000.
  const weightSum = proposal.constituents.reduce((sum, c) => sum + c.weightBps, 0);
  if (weightSum !== 10_000) {
    return {
      valid: false,
      error: `Weight sum is ${weightSum}, must be exactly 10000`,
    };
  }

  return { valid: true, proposal };
}

// ── Anthropic API call with one retry ────────────────────────────────────────

async function callAnthropicWithRetry(
  thesis: string,
  catalogue: CatalogueAsset[]
): Promise<AIProposal> {
  const systemPrompt = buildSystemPrompt();
  const userPrompt = buildUserPrompt(thesis, catalogue);

  async function attemptCall(isRetry: boolean): Promise<AIProposal> {
    const retryNote = isRetry
      ? "\n\nIMPORTANT: Your previous response was not valid JSON or did not meet the schema requirements. Return ONLY the raw JSON object with no markdown, no backticks, no explanation."
      : "";

    const message = await anthropic.messages.create({
      model: "claude-sonnet-4-20250514",
      max_tokens: 1000,
      system: systemPrompt,
      messages: [
        {
          role: "user",
          content: userPrompt + retryNote,
        },
      ],
    });

    const textBlock = message.content.find((b) => b.type === "text");
    if (!textBlock || textBlock.type !== "text") {
      throw new Error("No text content in Anthropic response");
    }

    const rawText = textBlock.text.trim();

    // Strip markdown code fences if the model wrapped the JSON.
    const jsonText = rawText
      .replace(/^```(?:json)?\s*/i, "")
      .replace(/\s*```$/, "")
      .trim();

    let parsed: unknown;
    try {
      parsed = JSON.parse(jsonText);
    } catch {
      throw new Error(`Response is not valid JSON: ${rawText.slice(0, 200)}`);
    }

    const validation = validateProposal(parsed, catalogue);
    if (!validation.valid) {
      throw new Error(validation.error);
    }

    return validation.proposal;
  }

  try {
    return await attemptCall(false);
  } catch (firstError) {
    console.warn(
      `AI composition first attempt failed: ${firstError instanceof Error ? firstError.message : firstError}. Retrying...`
    );
    try {
      return await attemptCall(true);
    } catch (secondError) {
      throw new Error(
        `AI composition failed after retry: ${secondError instanceof Error ? secondError.message : secondError}`
      );
    }
  }
}

// ── Catalogue fetch from Go backend ──────────────────────────────────────────

async function fetchCatalogue(): Promise<CatalogueAsset[]> {
  const backendURL = process.env.BACKEND_URL ?? "http://localhost:8080";

  const response = await fetch(`${backendURL}/catalogue`);
  if (!response.ok) {
    throw new Error(
      `Failed to fetch catalogue from backend: ${response.status} ${response.statusText}`
    );
  }

  const raw = await response.json();
  const parsed = z.array(CatalogueAssetSchema).safeParse(raw);
  if (!parsed.success) {
    throw new Error(`Invalid catalogue response from backend: ${parsed.error.message}`);
  }

  // Only pass active assets to the AI — inactive ones can't be used in baskets.
  return parsed.data.filter((a) => a.isActive !== false);
}

// ── Express server ────────────────────────────────────────────────────────────

const app = express();
app.use(express.json({ limit: "16kb" }));

app.post("/compose", async (req: Request, res: Response): Promise<void> => {
  const bodyParsed = ComposeRequestSchema.safeParse(req.body);
  if (!bodyParsed.success) {
    res.status(400).json({
      error: "Invalid request body",
      details: bodyParsed.error.message,
    });
    return;
  }

  const { thesis } = bodyParsed.data;

  try {
    const catalogue = await fetchCatalogue();

    if (catalogue.length < 3) {
      res.status(503).json({
        error: "Insufficient assets in catalogue to construct a basket (minimum 3 required)",
      });
      return;
    }

    const proposal = await callAnthropicWithRetry(thesis, catalogue);

    // Enrich the proposal with name and sector from the catalogue
    // so the frontend doesn't need a second lookup.
    const catalogueByAddress = new Map(
      catalogue.map((a) => [a.address.toLowerCase(), a])
    );

    const enriched = {
      constituents: proposal.constituents.map((c) => {
        const meta = catalogueByAddress.get(c.address.toLowerCase());
        return {
          address: c.address,
          symbol: c.symbol,
          name: meta?.name ?? "",
          sector: meta?.sector ?? "",
          weightBps: c.weightBps,
          rationale: c.rationale,
          currentPriceUsdg: meta?.currentPriceUsdg ?? "0",
        };
      }),
      overallRationale: proposal.overallRationale,
      riskNotes: proposal.riskNotes,
    };

    res.status(200).json(enriched);
  } catch (error) {
    const message = error instanceof Error ? error.message : "Unknown error";
    console.error(`/compose error: ${message}`);
    res.status(502).json({ error: message });
  }
});

app.get("/health", (_req: Request, res: Response): void => {
  res.status(200).json({ status: "ok", service: "weave-ai-agent" });
});

// ── Start ─────────────────────────────────────────────────────────────────────

const PORT = parseInt(process.env.AI_SERVICE_PORT ?? "3001", 10);

app.listen(PORT, () => {
  console.log(`Weave AI composition engine listening on :${PORT}`);
});

export { callAnthropicWithRetry, validateProposal, fetchCatalogue };

File: server/ai/compose.go
package ai

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
)

// CatalogueAsset is the shape the AI composer receives per stock.
type CatalogueAsset struct {
	Address          string `json:"address"`
	Symbol           string `json:"symbol"`
	Name             string `json:"name"`
	Sector           string `json:"sector"`
	CurrentPriceUsdg string `json:"currentPriceUsdg"`
}

// ProposedConstituent is one stock in the AI's returned basket proposal.
type ProposedConstituent struct {
	Address          string `json:"address"`
	Symbol           string `json:"symbol"`
	Name             string `json:"name"`
	Sector           string `json:"sector"`
	WeightBps        int    `json:"weightBps"`
	Rationale        string `json:"rationale"`
	CurrentPriceUsdg string `json:"currentPriceUsdg"`
}

// Proposal is the full AI composition response returned to the caller.
type Proposal struct {
	Constituents     []ProposedConstituent `json:"constituents"`
	OverallRationale string                `json:"overallRationale"`
	RiskNotes        string                `json:"riskNotes"`
	Provider         string                `json:"provider"`
}

// Composer calls the OpenAI chat completions API and validates the response.
type Composer struct {
	apiKey string
	model  string
}

func NewComposer(apiKey, model string) *Composer {
	return &Composer{apiKey: apiKey, model: model}
}

// Compose sends the thesis and catalogue to OpenAI and returns a validated proposal.
// Retries once with a stricter prompt if the first response fails validation.
func (c *Composer) Compose(ctx context.Context, thesis string, catalogue []CatalogueAsset) (*Proposal, error) {
	raw, err := c.call(ctx, systemPrompt(), userPrompt(thesis, catalogue))
	if err != nil {
		return nil, fmt.Errorf("openai call: %w", err)
	}

	proposal, err := stripAndValidate(raw, catalogue)
	if err != nil {
		// Retry once with a stricter JSON-only reminder.
		raw, err = c.call(ctx, systemPrompt(), userPrompt(thesis, catalogue)+retryNote())
		if err != nil {
			return nil, fmt.Errorf("openai retry call: %w", err)
		}
		proposal, err = stripAndValidate(raw, catalogue)
		if err != nil {
			return nil, fmt.Errorf("validation failed after retry: %w", err)
		}
	}

	// Enrich with name, sector, price from catalogue.
	byAddr := make(map[string]CatalogueAsset, len(catalogue))
	for _, a := range catalogue {
		byAddr[strings.ToLower(a.Address)] = a
	}

	for i, c := range proposal.Constituents {
		if meta, ok := byAddr[strings.ToLower(c.Address)]; ok {
			proposal.Constituents[i].Name             = meta.Name
			proposal.Constituents[i].Sector           = meta.Sector
			proposal.Constituents[i].CurrentPriceUsdg = meta.CurrentPriceUsdg
		}
	}

	proposal.Provider = "openai/" + c.model
	return proposal, nil
}

// openAIRequest is the request body shape for /v1/chat/completions.
type openAIRequest struct {
	Model    string              `json:"model"`
	Messages []openAIMessage     `json:"messages"`
	ResponseFormat *responseFormat `json:"response_format,omitempty"`
}

type openAIMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type responseFormat struct {
	Type string `json:"type"`
}

type openAIResponse struct {
	Choices []struct {
		Message struct {
			Content string `json:"content"`
		} `json:"message"`
	} `json:"choices"`
	Error *struct {
		Message string `json:"message"`
	} `json:"error"`
}

func (c *Composer) call(ctx context.Context, system, user string) (string, error) {
	body, err := json.Marshal(openAIRequest{
		Model: c.model,
		Messages: []openAIMessage{
			{Role: "system", Content: system},
			{Role: "user",   Content: user},
		},
		ResponseFormat: &responseFormat{Type: "json_object"},
	})
	if err != nil {
		return "", err
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost,
		"https://api.openai.com/v1/chat/completions",
		bytes.NewReader(body),
	)
	if err != nil {
		return "", err
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+c.apiKey)

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", fmt.Errorf("http: %w", err)
	}
	defer resp.Body.Close()

	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("read body: %w", err)
	}

	var oResp openAIResponse
	if err := json.Unmarshal(raw, &oResp); err != nil {
		return "", fmt.Errorf("unmarshal: %w", err)
	}

	if oResp.Error != nil {
		return "", fmt.Errorf("openai error: %s", oResp.Error.Message)
	}

	if len(oResp.Choices) == 0 || oResp.Choices[0].Message.Content == "" {
		return "", fmt.Errorf("empty response from openai")
	}

	return oResp.Choices[0].Message.Content, nil
}

func systemPrompt() string {
	return `You are a financial analyst building thematic equity baskets from a specific catalogue of tokenized stocks. You will receive a natural language investment thesis and a JSON catalogue of available stocks with their symbols, names, sectors, and current prices.

Your task is to select 3 to 12 stocks from the catalogue that best represent the thesis, assign each a target weight in basis points that sum to exactly 10,000, and provide a one-sentence rationale for each inclusion.

Rules:
- Only select stocks that appear in the provided catalogue
- Weights must be integers and must sum to exactly 10,000
- No single weight may be less than 100 (1%) or more than 5000 (50%)
- Do not include more than 12 constituents
- Do not include fewer than 3 constituents
- Prefer direct plays over indirect beneficiaries unless the thesis specifically calls for breadth
- Weight by conviction and relevance to the thesis, not by market cap alone

Return ONLY valid JSON matching this schema, no preamble or explanation outside the JSON:
{
  "constituents": [
    {
      "address": "0x...",
      "symbol": "AAPL",
      "weightBps": 2000,
      "rationale": "one sentence explaining why this stock fits the thesis"
    }
  ],
  "overallRationale": "two to three sentences explaining the basket overall construction logic",
  "riskNotes": "one to two sentences noting the key risks or concentration exposures"
}`
}

func userPrompt(thesis string, catalogue []CatalogueAsset) string {
	catJSON, _ := json.MarshalIndent(catalogue, "", "  ")
	return fmt.Sprintf("Investment thesis: %s\n\nAvailable stock catalogue:\n%s\n\nReturn only the JSON proposal with no additional text.", thesis, catJSON)
}

func retryNote() string {
	return "\n\nIMPORTANT: Your previous response was not valid JSON or did not meet schema requirements. Return ONLY the raw JSON object. No markdown, no backticks, no explanation."
}

// rawProposal is the shape we parse from the LLM before validation.
type rawProposal struct {
	Constituents []struct {
		Address   string `json:"address"`
		Symbol    string `json:"symbol"`
		WeightBps int    `json:"weightBps"`
		Rationale string `json:"rationale"`
	} `json:"constituents"`
	OverallRationale string `json:"overallRationale"`
	RiskNotes        string `json:"riskNotes"`
}

// stripAndValidate strips markdown fences, parses JSON, and validates
// all business rules.
func stripAndValidate(raw string, catalogue []CatalogueAsset) (*Proposal, error) {
	clean := raw
	clean = strings.TrimPrefix(clean, "```json")
	clean = strings.TrimPrefix(clean, "```")
	clean = strings.TrimSuffix(clean, "```")
	clean = strings.Trim(clean, `"`)
	clean = strings.ReplaceAll(clean, `\"`, `"`)
	clean = strings.TrimSpace(clean)

	var rp rawProposal
	if err := json.Unmarshal([]byte(clean), &rp); err != nil {
		return nil, fmt.Errorf("json parse: %w", err)
	}

	if len(rp.Constituents) < 3 {
		return nil, fmt.Errorf("too few constituents: %d (minimum 3)", len(rp.Constituents))
	}
	if len(rp.Constituents) > 12 {
		return nil, fmt.Errorf("too many constituents: %d (maximum 12)", len(rp.Constituents))
	}
	if strings.TrimSpace(rp.OverallRationale) == "" {
		return nil, fmt.Errorf("overallRationale is empty")
	}
	if strings.TrimSpace(rp.RiskNotes) == "" {
		return nil, fmt.Errorf("riskNotes is empty")
	}

	byAddr := make(map[string]CatalogueAsset, len(catalogue))
	for _, a := range catalogue {
		byAddr[strings.ToLower(a.Address)] = a
	}

	seen    := make(map[string]bool, len(rp.Constituents))
	weightSum := 0

	for _, c := range rp.Constituents {
		addr := strings.ToLower(c.Address)

		meta, ok := byAddr[addr]
		if !ok {
			return nil, fmt.Errorf("address %s (%s) not in catalogue", c.Address, c.Symbol)
		}
		if meta.Symbol != c.Symbol {
			return nil, fmt.Errorf("symbol mismatch at %s: catalogue has %s, proposal has %s", c.Address, meta.Symbol, c.Symbol)
		}
		if seen[addr] {
			return nil, fmt.Errorf("duplicate constituent: %s", c.Address)
		}
		seen[addr] = true

		if c.WeightBps < 100 {
			return nil, fmt.Errorf("weight too low for %s: %d bps (minimum 100)", c.Symbol, c.WeightBps)
		}
		if c.WeightBps > 5000 {
			return nil, fmt.Errorf("weight too high for %s: %d bps (maximum 5000)", c.Symbol, c.WeightBps)
		}
		if strings.TrimSpace(c.Rationale) == "" {
			return nil, fmt.Errorf("empty rationale for %s", c.Symbol)
		}

		weightSum += c.WeightBps
	}

	if weightSum != 10000 {
		return nil, fmt.Errorf("weight sum is %d, must be exactly 10000", weightSum)
	}

	constituents := make([]ProposedConstituent, len(rp.Constituents))
	for i, c := range rp.Constituents {
		constituents[i] = ProposedConstituent{
			Address:   c.Address,
			Symbol:    c.Symbol,
			WeightBps: c.WeightBps,
			Rationale: c.Rationale,
		}
	}

	return &Proposal{
		Constituents:     constituents,
		OverallRationale: rp.OverallRationale,
		RiskNotes:        rp.RiskNotes,
	}, nil
}